src/_avif.c

```diff
@@ -0,0 +1,819 @@
+            image->colorPrimaries = AVIF_COLOR_PRIMARIES_BT709;
+            image->transferCharacteristics = AVIF_TRANSFER_CHARACTERISTICS_SRGB;
+        }
+
+        if (exif_size) {
+            avifImageSetMetadataExif(image, exif_bytes, exif_size);
+        }
+        if (xmp_size) {
+            avifImageSetMetadataXMP(image, xmp_bytes, xmp_size);
+        }
+
+        self->image = image;
+        self->frame = NULL;
+
+        return (PyObject *)self;
+    }
+    PyErr_SetString(PyExc_RuntimeError, "could not create encoder object");
+    return NULL;
+}
+
+PyObject *
+_encoder_dealloc(AvifEncoderObject *self) {
+    if (self->encoder) {
+        avifEncoderDestroy(self->encoder);
+    }
+    if (self->image) {
+        avifImageDestroy(self->image);
+    }
+    if (self->frame) {
+        avifImageDestroy(self->frame);
+    }
+    Py_RETURN_NONE;
+}
+
+PyObject *
+_encoder_add(AvifEncoderObject *self, PyObject *args) {
+    uint8_t *rgb_bytes;
+    Py_ssize_t size;
+    int duration;
+    int width;
+    int height;
+    char *mode;
+    PyObject *is_single_frame = NULL;
+
+    int channels;
+    avifRGBImage rgb;
+    avifResult result;
+
+    avifEncoder *encoder = self->encoder;
+    avifImage *image = self->image;
+    avifImage *frame;
+
+    if (!PyArg_ParseTuple(
+            args,
+            "z#iiisO",
+            (char **)&rgb_bytes,
+            &size,
+            &duration,
+            &width,
+            &height,
+            &mode,
+            &is_single_frame)) {
+        return NULL;
+    }
+
+    if (image->yuvRowBytes[0] == 0) {
+        // If we don't have an image populated with yuv planes, this is the first frame
+        frame = image;
+    } else {
+        if (self->frame) {
+            avifImageDestroy(self->frame);
+        }
+        self->frame = avifImageCreateEmpty();
+
+        frame = self->frame;
+
+        frame->colorPrimaries = image->colorPrimaries;
+        frame->transferCharacteristics = image->transferCharacteristics;
+        frame->matrixCoefficients = image->matrixCoefficients;
+        frame->yuvRange = image->yuvRange;
+        frame->yuvFormat = image->yuvFormat;
+        frame->depth = image->depth;
+    }
+
+    frame->width = width;
+    frame->height = height;
+
+    if ((image->width != frame->width) || (image->height != frame->height)) {
+        PyErr_Format(
+            PyExc_ValueError,
+            "Image sequence dimensions mismatch, %ux%u != %ux%u",
+            image->width,
+            image->height,
+            frame->width,
+            frame->height);
+        return NULL;
+    }
+
+    memset(&rgb, 0, sizeof(avifRGBImage));
+
+    avifRGBImageSetDefaults(&rgb, frame);
+    rgb.depth = 8;
+
+    if (strcmp(mode, "RGBA") == 0) {
+        rgb.format = AVIF_RGB_FORMAT_RGBA;
+        channels = 4;
+    } else {
+        rgb.format = AVIF_RGB_FORMAT_RGB;
+        channels = 3;
+    }
+
+    avifRGBImageAllocatePixels(&rgb);
+    memcpy(rgb.pixels, rgb_bytes, size);
```

Comment:
Please document in a comment that this is safe for r/w, and potentially add an explict check that the rgb_bytes/rgb.pixels is large enough.